src/hera/workflows/script.py

```diff
@@ -383,14 +383,14 @@ def append_annotation(annotation: Union[Artifact, Parameter]):
             parameters.append(annotation)
 
     return_annotation = inspect.signature(source).return_annotation
-    if get_origin(return_annotation) is Annotated:
-        append_annotation(get_args(return_annotation)[1])
+    if param_or_artifact := _type_util.get_annotated_metadata(return_annotation, (Artifact, Parameter)):
+        append_annotation(param_or_artifact)
     elif get_origin(return_annotation) is tuple:
         for annotation in get_args(return_annotation):
             if isinstance(annotation, type) and issubclass(annotation, (OutputV1, OutputV2)):
                 raise ValueError("Output cannot be part of a tuple output")
 
-            append_annotation(get_args(annotation)[1])
+            append_annotation(_type_util.get_annotated_metadata(annotation, (Artifact, Parameter)))
```

Comment:
We need to revisit this code because of #1167 anyway, so I would suggest the same `if param_or_artifact :=` pattern here, and remove the `None` from the `append_annotation` union:

```suggestion
            if param_or_artifact := _type_util.get_annotated_metadata(return_annotation, (Artifact, Parameter)):
                append_annotation(param_or_artifact)
```

Reply:
IMO the refactor has already incorporated a lot of implicit assumption fixes as described in https://github.com/argoproj-labs/hera/pull/1168/files#r1734338480 which would have to be undone and verified.

We've simplified a lot of the `if X is Annotated` checks which were implicitly tying up `if Annotated` with the assumption that the annotation is a `Parameter` or `Artifact`. I doubt users have `Annotated` with non-Hera metadata in the context of Hera if they want their code to work as expected, plus it's still an experimental feature so no need to be super careful.

I think it would be simpler to fix everything at once as it would be harder to guarantee it as a strict refactor (and I'm not fussed about it needing to be a strict refactor).